baseapp/abci.go

```diff
@@ -248,11 +248,29 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) (resp abci.ResponsePrepareProposal) {
-	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
 	if app.prepareProposal == nil {
 		panic("PrepareProposal method not set")
 	}
 
+	// Tendermint must never call PrepareProposal with a height of 0.
+	// Ref: https://github.com/tendermint/tendermint/blob/059798a4f5b0c9f52aa8655fa619054a0154088c/spec/core/state.md?plain=1#L37-L38
+	if req.Height < 1 {
+		panic("PrepareProposal called with invalid height")
+	}
+
+	ctx := app.prepareProposalState.ctx
+	// Here we use deliverState on the first block given that we want to be able
```

Comment:
Curious - why is the state with changes that InitChain made not available to the app?

Reply:
wdym "not available to the app"?

After `InitChain`, changes are not committed. They're technically staged/cached on the (cached) root multi-store.

Reply:
InitChain initializes and applies changes (params, genesis and such) to `deliverState`, which is not persisted until Commit(): [see here](https://github.com/cosmos/cosmos-sdk/blob/6ac0c3628e5786614dd6bb3183b2652e7486cd7f/baseapp/abci.go#L402-L405).
In `Commit()` `deliverState` is set to nil while the other states are set to the latest header (that's why we were seeing the previous block height in `PrepareProposal` before adding `WithBlockHeight(req.Height)` in this PR).

So `PrepareProposal` and `ProcessProposal` will always be seeing persisted state (from the previous produced block), except for the first block that we feed it with the genesis state. Also it's good to note that we shouldn't be making any persistable state changes during these stages.

Also `BeginBlock` is the one in charge of setting deliverState; except when we are at the first block: in that case we just want to update the context retaining any state changes made in `InitChain`. [see here](https://github.com/cosmos/cosmos-sdk/blob/6ac0c3628e5786614dd6bb3183b2652e7486cd7f/baseapp/abci.go#L161-L172)

Reply:
I see, thanks for the explanation. That makes sense.

Random alternative thoughts: 
1. Would it be better to dedicate height == 1 to `InitChain` only?  meaning that the at height 1, CosmosSDK just runs `InitChain` and then calls `Commit`. No special logic to override the ctx state to be the `deliverState` which has not been committed. My main reasoning around this is if there would be any weird behavior for the `PrepareProposal` to be based off of state that has not been committed to state.
2. Similarly, is it possible to introduce some special height 0 that does the `InitChain` and `Commit` only